A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,207 @@
+    - If the address is ejected, and the current time is after `ejection_timestamp + min(base_ejection_time * multiplier, max(base_ejection_time, max_ejection_time))`, un-eject the address by setting the ejection timestmap to `null` and calling `uneject` on each subchannel wrapper.
+
+### Call Counter
+
+This design is based directly on Envoy's implementation. The object contains two buckets, and each bucket has a number counting successes, and another counting failures. The active bucket is updated each time a call finishes. When the timer triggers, the inactive bucket is zeroed and swapped with the active bucket. Then the inactive bucket contains the number of successes and failures since the last time the timer triggered. Those numbers are used to evaluate the ejection criteria.
+
+### Subchannel Wrapper
+
+The subchannel wrappers created by the `outlier_detection` LB policy will hold a reference to its map entry in the LB policy, if that map entry exists.
+
+The subchannel wrapper will track the latest state update from the underlying subchannel. By default, it will simply pass those updates along. The wrapper will have the following methods:
+
+ - `eject()`: The wrapper will report a state update with the `TRANSIENT_FAILURE` state, and will stop passing along updates from the underlying subchannel.
+ - `uneject()`: The wrapper will report a state update with the latest update from the underlying subchannel, and resume passing along updates from the underlying subchannel.
+
+### Success Rate Algorithm
+
+ 1. If the number of addresses with request volume of at least `success_rate_request_volume` is less than `success_rate_minimum_hosts`, stop.
+ 2. Calculate the mean and standard deviation of the fractions of successful requests among addresses with total request volume of at least `success_rate_request_volume`.
+ 3. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `success_rate_request_volume`, continue to the next address.
+    3. If the address's success rate is less than `(mean - stdev * (success_rate_stdev_factor / 1000))`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_success_rate`, eject that address.
+
+### Failure Percentage Algorithm
+
+ 1. If the number of addresses is less than `failure_percentage_minimum_hosts`, stop.
+ 2. For each address:
+    1. If the percentage of ejected addresses is greater than `max_ejection_percent`, stop.
+    2. If the address's total request volume is less than `failure_percentage_request_volume`, continue to the next address.
+    3. If the address's failure percentage is greater than `failure_percentage_threshold`, then choose a random integer in `[0, 100)`. If that number is less than `enforcing_failure_percentage`, eject that address.
+
+### xDS Integration
```

Comment:
In this section, we should be very specific about how each of the fields in the xDS outlier detection config will be handled by gRPC.  In particular:
- We should explicitly list all of the xDS fields that we will look at, what values are legal, and how we will handle any illegal values in each field -- e.g., will we NACK or ignore the value or what?
- We should be explicit about which fields in the xDS config we will *not* support, and how we will handle those fields.  In the case of this particular design, I suspect we can just say something like "any field not listed above will be ignored by gRPC".

We need to pay careful attention to forward-compatibility here, especially in cases where we might decide to NACK.  We should carefully think through what would happen if in the future we add support for new fields and make sure that the behavior that we start with will not cause any problems.

Reply:
One concern I have is with the "Consecutive 5xx" ejection algorithm, which is enabled by default in the xDS configuration. If a user does not explicitly configure it, then the behavior will still change if we add support for it. I don't know if that's a significant problem, and it's also unlikely that we will ever support it.

Reply:
I think that particular case is probably okay.  As you say, we're unlikely to ever support that.  And even if we did, (a) there's no requirement that gRPC has the same defaults as Envoy, and (b) we could probably even get away with enabling this by default, since it takes effect only when the backends are failing requests anyway.  So we'd have some options here.